src/main/java/org/wildfly/security/x500/cert/ocsp/OcspChainVerifier.java

```diff
@@ -0,0 +1,118 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.x500.cert.ocsp;
+
+import static org.wildfly.security.asn1.util.ASN1.CONTEXT_SPECIFIC_MASK;
+import static org.wildfly.security.x500.GeneralName.URI_NAME;
+import static org.wildfly.security.x500.X500.OID_AD_OCSP;
+
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.util.Collection;
+
+import org.wildfly.security._private.ElytronMessages;
+import org.wildfly.security.asn1.DERDecoder;
+import org.wildfly.security.x500.GeneralName;
+import org.wildfly.security.x500.X500;
+
+/**
+ * The certificate chain verifier.
+ */
+public class OcspChainVerifier {
+
+    private final URL responder;
+    private final OcspSignatureVerifier signatureVerifier;
+    private final OcspChecker checker;
+    private final boolean acceptUnknownCertificates;
+    private final boolean acceptWhenStatusUnavailable;
+
+    /**
+     * Construct the certificate chain verifier.
+     *
+     * @param checker the checker providing status of individual certificates
+     * @param responder the responder to use to check all certificates ({@code null} to use responder from the certificate)
+     * @param trusted the collection of trusted OCSP responders certificates
+     * @param acceptUnknownCertificates whether should be certificate unknown by OCSP responder considered valid
+     * @param acceptWhenStatusUnavailable whether should be certificate considered valid when OCSP responder is unavailable
+     */
+    public OcspChainVerifier(OcspChecker checker, URL responder, Collection<X509Certificate> trusted, boolean acceptUnknownCertificates, boolean acceptWhenStatusUnavailable) {
+        this.checker = checker;
+        this.responder = responder;
+        this.signatureVerifier = new OcspSignatureVerifier(trusted);
+        this.acceptUnknownCertificates = acceptUnknownCertificates;
+        this.acceptWhenStatusUnavailable = acceptWhenStatusUnavailable;
+    }
+
+    /**
+     * Check the certificate chain.
+     *
+     * @param chain the certificate chain
+     * @throws CertificateException if the validation fails (some certificate in the chain is revoked for example)
+     */
+    public void checkChain(X509Certificate[] chain) throws CertificateException {
+        for (int i = 0; i < chain.length - 1; i++) {
+            X509Certificate certificate = chain[i];
+            X509Certificate issuer = chain[i + 1];
```

Comment:
Possible ArrayIndexOutOfBoundsException?

Reply:
There is `i < chain.length - 1`